### PR TITLE
Tag mapping in graph refresh

### PR DESCRIPTION
**Pull request by cben**
Depends on:
- [x] https://github.com/ManageIQ/manageiq/pull/16453 for specs
- [x] https://github.com/ManageIQ/manageiq/pull/16499 for fixing that spec
- [x] https://github.com/ManageIQ/manageiq/pull/16501 to define `Tag.controlled_by_mapping` scope — merged
    (~~OR revive reverted https://github.com/ManageIQ/manageiq/pull/16449~~)

Overview/plan issue: #126

(added test for not removing user-assigned tag has been extracted to #170 and already merged)

@miq-bot add-label enhancement gaprindashvili/yes
https://bugzilla.redhat.com/show_bug.cgi?id=1470021

@Ladas @moolitayer please review

### CODE REVIEW

**Comment by cben** [comment]
@miq-bot add-label enhancement, gaprindashvili/yes

File: app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb [hunk 1]
```diff
@@ -355,6 +368,32 @@ def initialize_custom_attributes_collections(parent_collection, sections)
     end
   end
 
+  def initialize_taggings_collection(parent_collection)
+    type = parent_collection.model_class.base_class.name
+    relation = parent_collection.full_collection_for_comparison
+    # TODO: this will remove user-assigned tags!
```
Comment (Ladas) [inline app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb]: do we delete user assigned tags in old refresh?

Comment (cben) [reply app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb]: no, `retag_entity` was ignoring those, using `controls_tag?` which did several queries per entity per tag.
https://github.com/ManageIQ/manageiq/pull/16449 replaces that with a (scary) join scope, updated this PR to use that.
it's unproven, working on adding tests...

**Comment by cben** [comment]
@moolitayer please review.  rebased/squashed.
Core PRs are merged.  I'm afraid this repo is now red :red_circle:, first commit here fixes (and improves) tests, you might want to merge that first :green_apple:.  Second commit adds tag mapping in graph refresh :cake:.

**Comment by cben** [comment]
@moolitayer The combination of the 2 core PRs broke travis, debugging but I might need to roll the scope PR back :(
So review but don't merge this yet.

**Comment by bazulay** [comment]
@moolitayer PTAL

File: app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb [hunk 2]
```diff
@@ -256,6 +264,8 @@ def initialize_inventory_collections
         )
       )
     initialize_custom_attributes_collections(@collections[:container_templates], %w(labels))
+    initialize_taggings_collection(@collections[:container_templates])
```
Comment (moolitayer) [inline app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb]: I seem to be missing the refresh_parser.rb code that handles template as well as build_pods for that matter, where is it?

File: app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb [hunk 3]
```diff
@@ -291,6 +303,7 @@ def initialize_inventory_collections
         )
       )
     initialize_custom_attributes_collections(@collections[:container_build_pods], %w(labels))
+    initialize_taggings_collection(@collections[:container_build_pods])
```
Comment (moolitayer) [inline app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb]: seems like we can remove this one

Comment (cben) [reply app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb]: Yes, removed (ContainerBuildPod model doesn't acts_as_miq_taggable, and don't have their own page in UI where one could see/edit tags.)

As you noticed, templates and images are also missing `map_labels`, that's a bug I'll fix separately but I prefer to keep their taggings_for_ collections in these PR.

File: app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb (hunk 2 above)
Comment (cben) [reply app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb]: As we discussed offline:
- removed ContainerBuildPod (that model doesn't acts_as_miq_taggable, and don't have their own page in UI where one could see/edit tags).
- indeed templates and images are also missing `map_labels()`, that's a bug I'll fix separately but I prefer to keep their `taggings_for_` collections in these PR.

PTAL.

File: app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb [hunk 4]
```diff
@@ -355,6 +368,31 @@ def initialize_custom_attributes_collections(parent_collection, sections)
     end
   end
 
+  def initialize_taggings_collection(parent_collection)
+    type = parent_collection.model_class.base_class.name
+    relation = parent_collection.full_collection_for_comparison
+    # TODO: test this.
```
Comment (moolitayer) [inline app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb]: is this comment for manual or automated testing?
any change can we remove it already?

Comment (cben) [reply app/models/manageiq/providers/kubernetes/container_manager/inventory_collections.rb]: tbh, I don't remember.  removed.
I'm OK with existing check in refresher_spec (extended in #170), and I'll add checks to cover each entity next time I re-record the VCR.

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit cee61f60877dbc7a07cfca7330ce94a2a64cb391
Author: Mooli Tayer <mtayer@redhat.com>
Date:   Sun Dec 3 10:37:24 2017 +0200

    Merge pull request #162 from cben/tag-mapper-InventoryCollection
    
    Tag mapping in graph refresh
    (cherry picked from commit f58659bc6bb6a13c72df817b0626184b96829518)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1520488
```

**Pull request merged**